pkg/agent/controller/traceflow/traceflow_controller.go

```diff
@@ -0,0 +1,386 @@
+		c.queue.Forget(obj)
+		klog.Errorf("Expected string in work queue but got %#v", obj)
+		return true
+	} else if err := c.syncTraceflow(key); err == nil {
+		// If no error occurs we Forget this item so it does not get queued again.
+		c.queue.Forget(key)
+	} else {
+		// If error occurs we log error.
+		klog.Errorf("Error syncing Traceflow %s, Aborting. Error: %v", key, err)
+	}
+	return true
+}
+
+// syncTraceflow gets Traceflow CRD by name, update cache and start syncing.
+func (c *Controller) syncTraceflow(traceflowName string) error {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished syncing Traceflow for %s. (%v)", traceflowName, time.Since(startTime))
+	}()
+
+	tf, err := c.traceflowLister.Get(traceflowName)
+	if err != nil {
+		return err
+	}
+	switch tf.Status.Phase {
+	case opsv1alpha1.Running:
+		if tf.Status.DataplaneTag != 0 {
+			if _, ok := c.running[tf.Status.DataplaneTag]; !ok {
+				c.running[tf.Status.DataplaneTag] = *tf
+				err = c.startTraceflow(tf)
+			} else {
+				c.UpdateTagCache(tf)
+			}
+		} else {
+			klog.Warningf("Invalid data plane tag %d for Traceflow %s", tf.Status.DataplaneTag, tf.Name)
+		}
+	default:
+		c.deallocateTag(tf)
+	}
+	return err
+}
+
+// startTraceflow deploys OVS flow entries for Traceflow and inject packet if current Node
+// is Sender Node.
+func (c *Controller) startTraceflow(tf *opsv1alpha1.Traceflow) error {
+	// Deploy flow entries for traceflow
+	klog.V(2).Infof("Deploy flow entries for Traceflow %s", tf.Name)
+	err := c.ofClient.InstallTraceflowFlows(tf.Status.DataplaneTag)
+	defer func() {
+		if err != nil {
+			c.errorTraceflowCRD(tf)
+		}
+	}()
+	if err != nil {
+		return err
+	}
+
+	// Inject packet if this Node is sender.
+	time.Sleep(time.Duration(injectPacketDelay) * time.Second)
+	err = c.injectPacket(tf)
+	return err
+}
+
+func (c *Controller) injectPacket(tf *opsv1alpha1.Traceflow) error {
+	namespaceName := tf.Spec.Source.Namespace
+	podName := tf.Spec.Source.Pod
+	podInterfaces := c.interfaceStore.GetContainerInterfacesByPod(podName, namespaceName)
+	// Pod not found in current Node, skip inject packet.
+	if len(podInterfaces) == 0 {
+		return nil
+	}
+
+	// Update Traceflow phase to Running.
+	klog.V(2).Infof("Injecting packet for Traceflow %s", tf.Name)
+	c.senders[tf.Status.DataplaneTag] = *tf
+
+	// Calculate destination MAC/IP.
+	dstMAC := ""
+	dstIP := tf.Spec.Destination.IP
+	// TODO: Find MAC by dstIP
+	if dstIP == "" {
+		dstPodInterfaces := c.interfaceStore.GetContainerInterfacesByPod(tf.Spec.Destination.Pod, tf.Spec.Destination.Namespace)
+		if len(dstPodInterfaces) > 0 {
+			dstMAC = dstPodInterfaces[0].MAC.String()
+			dstIP = dstPodInterfaces[0].IP.String()
+		} else {
+			dstPod, err := c.kubeClient.CoreV1().Pods(tf.Spec.Destination.Namespace).Get(context.TODO(), tf.Spec.Destination.Pod, v1.GetOptions{})
+			if err != nil {
+				return err
+			}
+			// dstMAC is "" here, will be set to Gateway MAC in ofClient.SendTraceflowPacket
```

Comment:
I feel clearer to set gateway MAC in this func, but no strong opinion.

Reply:
I also do not have strong preference. Let's keep existing implementation in this PR.